[Feature] 배틀 결과 조회 API 구현하기

rhxfkg opened the pull request:
## 🔗 관련 이슈
<!-- 예: Closes #123, Relates to #456 -->

#10, Closes #13 

## ✅ 작업 내용

### 💡개선 사항
- 배틀 종료 후 결과 조회 API 구현 (`GET /battles/:id/result`)
- TDD 방식으로 테스트 코드 먼저 작성 후 구현 진행
- 3가지 배틀 시나리오 목데이터 생성 (A팀 승리, B팀 승리, 무승부)
- MVP 계산 로직 구현 (타임라인 upvotes 기반)
- 에러 핸들링 구현 (404: 존재하지 않는 배틀, 400: 진행 중인 배틀)
- CORS 설정 추가로 프론트엔드 연동 준비 완료

**주요 응답 데이터:**
- 배틀 기본 정보 (제목, 설명, 언어, 카테고리 등)
- 코드 A/B
- 투표 결과 (A팀/B팀/중립 득표수 및 비율)
- 통계 지표 (참여자 수, 조회수, 작전 전략 개수)
- 턴별 투표 추세 (누적값)
- 배틀 타임라인 (이의제기/반론만, 오래된순 정렬)
- MVP 정보 (최다 upvotes 획득자)

**구현 파일:**
- `battles.service.ts`: 비즈니스 로직 및 MVP 계산
- `battles.controller.ts`: API 엔드포인트 및 에러 핸들링
- `battles.module.ts`: NestJS 모듈 정의
- `battleResult.dto.ts`: 9개 DTO 클래스 정의
- `battleResults.mock.ts`: 4개 배틀 시나리오 목데이터
- `battles.service.spec.ts`: 8개 테스트 케이스
- `battles.controller.spec.ts`: 3개 테스트 케이스

**테스트 결과:**
- ✅ 13개 테스트 모두 통과 (2 test suites)
- ✅ curl 테스트 성공 (정상 케이스 3개, 에러 케이스 2개)

<br />

## 📸 참고 사항
<!-- 코드 리뷰에서 참고할 사항, 새로 주의 깊게 봐야 하는 파일/코드, 주요 고민과 해결 과정-->

### Unit Test
<img width="473" height="218" alt="image" src="https://github.com/user-attachments/assets/b36c144c-b41b-49fb-a553-02131430c8ba" />

### API 서버 실행 및 테스트
1. 성공 케이스 - A팀 승리
curl -s http://localhost:3000/battles/battle-1/result | jq
<details>
<summary>테스트 결과 이미지</summary>

<img width="1494" height="890" alt="image" src="https://github.com/user-attachments/assets/85dfdbfd-3f19-49a5-a936-c22ca1a3d3d4" />
<img width="708" height="909" alt="image" src="https://github.com/user-attachments/assets/0efa2983-eaf8-4ca5-9d82-84e1f9570780" />

</details>
<br />

2. 성공 케이스 - B팀 승리
curl -s http://localhost:3000/battles/battle-2/result | jq
<details>
<summary>테스트 결과 이미지</summary>

<img width="1488" height="909" alt="image" src="https://github.com/user-attachments/assets/1bb9612a-c2c9-4fe7-ae12-6ec2d4984cb7" />
<img width="711" height="905" alt="image" src="https://github.com/user-attachments/assets/f837bda6-36aa-47fe-a37c-3fe760140550" />

</details>
<br />

3. 성공 케이스 - 무승부
curl -s http://localhost:3000/battles/battle-3/result | jq
<details>
<summary>테스트 결과 이미지</summary>

<img width="1226" height="905" alt="image" src="https://github.com/user-attachments/assets/3a4d38ec-5f82-4ed6-aa65-71e6f68c7d41" />
<img width="632" height="882" alt="image" src="https://github.com/user-attachments/assets/6ce3d31b-1f40-4349-b4fd-708cb8f646d7" />

</details>
<br />

4. 404 에러 케이스
curl -s http://localhost:3000/battles/battle-999/result | jq
<details>
<summary>테스트 결과 이미지</summary>

<img width="605" height="106" alt="image" src="https://github.com/user-attachments/assets/10334759-ddb7-4da9-9fe3-53a16b408fd6" />

</details>
<br />

5. 400 에러 케이스 (진행 중인 배틀)
curl -s http://localhost:3000/battles/battle-open-1/result | jq
<details>
<summary>테스트 결과 이미지</summary>

<img width="628" height="103" alt="image" src="https://github.com/user-attachments/assets/43bd99f3-05c2-4ba7-9008-496c3e01957b" />

</details>

<br />

## 💬 질문사항 (고민했던 부분)
<!-- 리뷰어에게 묻고 싶은 것, 트레이드오프 등 -->

    1. 배틀 조회수
        1. 배틀을 참여하지 않더라도 들어온 경우도 포함?
        2. 그런데 애초에 진영을 선택하지 않으면 확인할 수가 없지 않나?
        3. 조회수 중복은 어떻게 할 것인가?
            1. 사용자별 조회 기록 저장 (로그인 필요하므로 X)
            2. IP 기반 (비로그인 가능)
                1. 실제 서비스에서 많이 사용하는 방식
                2. 같은 네트워크에서 여러 명 접속 시 1명으로 카운트될 수 있음
                3. DB저장 필요 현재 목데이터 환경이므로 redis에 저장하면 되긴하는데 redis를 그 목적으로 사용할 것이 아니지 않나
            3. 세션 기반 (프론트에서 처리)
                1. 구현 간단, DB저장 불필요
                2. 사용자가 localStorage 삭제 시 카운트됨 -> 조작 가능
    2. 투표 추세 그래프
        1. 단위를 어떻게 할 것인가
            1. 몇 분 기준으로? -> 게임 시간에 따라 다르게 할 것인지
            2. 턴에 맞춰서 할 것인지
    3. Mvp
        1. 타임라인에 있는 사용자들 중 누적해서 가장 많은 투표수를 받은 사람 (승패 상관X)
        2. 동일 수인 경우에는 먼저 의견 낸 사람

rhxfkg committed:
test: 배틀 결과 조회 테스트 코드 작성 #13

rhxfkg committed:
feat: 배틀 결과 목데이터 생성 #13

rhxfkg committed:
feat: 배틀 결과 DTO 정의 #13

rhxfkg committed:
feat: 배틀 결과 조회 API 구현 #13

rhxfkg committed:
refactor: 목데이터 사용으로 불필요한 async/await 제거 #13

rhxfkg committed:
chore: CORS 설정 추가 #13

rhxfkg committed:
style: lint-staged 포맷팅 적용 #13

jjjjjinseo commented on backend/src/battles/battles.service.ts:
```diff
@@ -0,0 +1,58 @@
+      status: 'CLOSED' as const,
+      mvp: mvp || battle.mvp,
+    }
+  }
```
DTO가 클래스로 잘 분리되어 있어서,
응답 객체를 DTO 쪽의 `of` / `fromEntity` 같은 정적 메서드를 통해 생성한다면 서비스 코드가 더 깔끔해질 것 같습니다!

rhxfkg replied:
레오 코드 참고해서 수정했습니다! 👍

jjjjjinseo commented on backend/src/battles/battles.service.ts:
```diff
@@ -0,0 +1,58 @@
+    return {
+      userId,
+      nickname: data.nickname,
+      team: data.team as 'A' | 'B',
```
'A' | 'B' 타입 단언은 공통 타입으로 분리하는건 어떨까요?!

rhxfkg replied:
같이 수정했습니다 👍

jjjjjinseo reviewed (approved):
수고하셨습니다! 테스트를 되게 꼼꼼하게 해주셨네요 👍 

고민하신 부분에 대해 생각해봤는데, 
제 생각엔 조회수는 정확도보다는 인기 지표라 중복을 허용해도 크게 문제는 없어 보입니다! 
굳이 중복 방지를 한다면 DB에 저장하는 구조보다 언급해주신 세션/브라우저 단위로 처리하는 방식이 좋은 것 같아요

투표 집계는 이전에 1분 단위로 정했던걸로 기억하는데(아마도), MVP 선정 단계에서도 간단하게 분 단위 집계로 가져가도 괜찮을 것 같습니다.!

HoonDongKang commented on backend/src/battles/dto/battleResult.dto.ts:
```diff
@@ -0,0 +1,69 @@
+export class VoteResultDto {
+  votes: number
+  percentage: number
+}
```
나중에 머지되고 `class-validator`를 설치되면 validation pipe로 검사해줘도 좋을 것 같아요~

HoonDongKang commented on backend/src/battles/controller/battles.controller.ts:
```diff
@@ -0,0 +1,21 @@
+export class BattlesController {
+  constructor(private readonly battlesService: BattlesService) {}
+
+  @Get(':id/result')
```
생각해보니 `/battles/:id/result` 엔드포인트를 명세서에 추가를 안해두었네요.. 
머지되고 노션에 기입 부탁드립니다 :)

U-S-jun replied:
상세 페이지도 엔드포인트를 나누는건가요? 저는 battles/:id 로 조회한다고 생각했어요
진행 중인 배틀의 경우 소켓 연결로 가고 아닐 경우(redis에 없는 배틀일 경우) DB 조회 후 결과 페이지로 이동하는 방식으로 생각하고있는데 스크럼때 논의해봅시다!

rhxfkg replied:
redis를 고려하지 못했네요

그러면 redis에서 진행중인지 아닌지를 먼저 확인하고
아니라면 끝난 배틀이니까 DB에서 조회해서 결과 페이지 렌더링
요롷게 하는게 좋아보이네요 👍

HoonDongKang commented on backend/src/battles/battles.controller.ts:
```diff
@@ -0,0 +1,21 @@
+    } catch (error) {
+      if (error instanceof NotFoundException || error instanceof BadRequestException) {
+        throw error
+      }
```
```suggestion
      if (error instanceof HttpException) {
        throw error
      }
```

Service에서 Exception이 추가될수록 해당 부분도 수정되니 유지보수성이 낮다고 생각이 듭니다. 
어차피 `HttpException`을 상속받으니, 이런 방식은 어떠신가요?

rhxfkg replied:
수정했습니다! 👍

HoonDongKang commented on backend/src/main.ts:
```diff
@@ -3,6 +3,12 @@ import { AppModule } from './app.module'
   const app = await NestFactory.create(AppModule)
+
+  app.enableCors({
+    origin: 'http://localhost:5173',
```
이 부분도 환경변수로 관리하면 좋을 것 같아요!

rhxfkg replied:
수정했습니다! 👍

HoonDongKang reviewed (approved):
고생하셨습니다!

개인적으로 Postman 이런 도구를 사용하지 않고 `curl`을 이용해서 터미널에서 요청보내서 테스트하신 게 인상적이네요! 따로 그렇게 요청보내신 이유가 있는 지 궁금합니다.

> 고민하신 부분

1. 프로토타입에서는 관리할 필요는 없다고 생각이 들고, 실제 서비스 개발 시작하면 유저 식별자로 제한해도 좋을 것 같습니다! (중립도 조회는 했으니 포함)
2. 투표는 어차피 하나의 Round가 마무리될 때마다 변동되니, 분 단위로 저장하는 것은 의미가 없을 것 같고 아마 `battle:roundupdate` 이벤트와 함께 데이터를 수집하지 않을까 싶습니다.
3. MVP는 한 번 고민해볼 필요가 있을 것 같아요 ㅋㅋㅋ 알고리즘이 따로 필요할 것 같습니다. 예를 들어, 가장 배틀에 오래 남아있는 사람 + 의견을 많이 낸 사람 + 항목에서 가장 많이 투표 받은 사람 등등으로 더해나가야하지 않을까...

rhxfkg committed:
refactor: fromEntity 패턴 적용 및 타입 안정성 개선 #13

- BattleResultResponseDto에 fromEntity 정적 메서드 추가
- Service에서 스프레드 연산자 대신 fromEntity 사용
- calculateMVP에서 타입 단언 제거 (Map 타입 명시)

rhxfkg committed:
refactor: HttpException 통합 및 CORS 환경변수화 #13

- Controller에서 HttpException으로 에러 처리 통합
- CORS origin을 환경변수로 관리 (FRONTEND_URL)
- .env.example 파일 추가 (환경변수 템플릿)

rhxfkg committed:
refactor: DTO와 Type 분리 #13

- types/battleResult.types.ts 생성하여 도메인 타입 정의
- DTO는 API 응답용 class만 유지
- Service에서 타입 import 경로 수정
- 관심사 분리 및 재사용성 향상

rhxfkg committed:
refactor: 폴더 구조 재구성 #13

- controller/, service/ 폴더로 분리
- mock 파일을 battles/mock/으로 이동
- import 경로 수정
- dev 브랜치 구조에 맞춰 정리

rhxfkg committed:
Merge branch 'dev' into feature/#13-battle-result-api

Signed-off-by: rhxfkg <51471056+rhxfkg@users.noreply.github.com>

rhxfkg committed:
refactor: dev 브랜치와 병합 준비 #13

- battles.service.ts: dev와 우리 브랜치 메서드 통합
- battles.controller.ts: 4개 엔드포인트 통합 (라우트 순서 조정)
- battles.controller.spec.ts: mock 기반 테스트로 통일
- battles.service.spec.ts: dev 방식으로 통일
- uuid 패키지 설치
- 중복 파일 제거 (battles.controller.spec.ts)
- 모든 테스트 통과 (22/22)
